java/languages.antlr/src/org/netbeans/modules/languages/antlr/v4/Antlr4Formatter.java

```diff
@@ -0,0 +1,149 @@
+        final int cstart = context.startOffset();
+        final int cend = context.endOffset();
+        
+        int tstart = token.getStartIndex();
+        int tstop = token.getStopIndex();
+        
+        try {
+            boolean inRule = false;
+            int ruleLine = 0;
+            int parenDepth = 0;
+            int textDelta = 0;
+            int prevTokenType = -1;
+            while ((token.getType() != EOF) && (tstart < cend)) {
+                switch (token.getType()) {
+                    case RULE_REF:
+                    case TOKEN_REF:
+                        // @header , @member, etc, are not real rules;
+                        if (!inRule && (prevTokenType != AT)) {
+                            inRule = true;
+                        }
+                        break;
+                    case LPAREN:
+                        parenDepth++;
+                        break;
+                    case RPAREN:
+                        parenDepth--;
+                        break;
+                    case SEMI:
+                        inRule = false;
+                        break;
+                }
+                if (tstop >= cstart) {
+                    if (!context.isIndent()) {
+                        //TODO: Do non-indent formatting
+                    }
+                    if (token.getChannel() == OFF_CHANNEL) {
+                        String ttext = token.getText();
+                        int nl = ttext.indexOf('\n');
+                        while ((nl != -1) && (tstart + nl <= cend))  {
+                            int lineStart = context.lineStartOffset(tstart + textDelta + nl);
+                            if (inRule || ruleLine > 0) {
+                                if (tstart + nl >= cstart) {
+                                    // Indent the first rule line to 0 the rest to indentSize
+                                    int originalIndent = context.lineIndent(lineStart);
+                                    int newIndent = ruleLine > 0 ? indentSize : 0;
+                                    context.modifyIndent(lineStart, newIndent);
+                                    textDelta += newIndent - originalIndent;
+                                }
+                                ruleLine = inRule ? ruleLine + 1 : 0;
+                            }
+                            nl = ttext.indexOf('\n', nl + 1);
+                        }
+                    }
+                }
+                prevTokenType = token.getType();
+                token = lexer.nextToken();
+                tstart = token.getStartIndex();
+                tstop = token.getStopIndex();
+            }
+        } catch (BadLocationException ex) {}
+    }
+
+    @Override
+    public void reindent(Context context) {
+        reformat(context, null);
+    }
+
+    @Override
+    public boolean needsParserResult() {
+        return false;
+    }
+
+    @Override
+    public int indentSize() {
+        return 3;
+    }
```

Comment:
3 space indent is evil since it looks almost right :)

Reply:
Just checked a few grammars today randomly. Most of them were using 4 spaces, but a few of them were using 3. It seems our ANTLR v3 and v4 grammars using 3 spaces are kind of a minority.